lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
 	DeleteToken(ctx context.Context, token string) error
 }
 
+type NodeJoinWait struct {
+	cfg *NodeJoinWaitConfig
+}
+
+func NewNodeJoinWait(cfg *NodeJoinWaitConfig) (*NodeJoinWait, error) {
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	return &NodeJoinWait{cfg: cfg}, nil
+}
+
+// Run grabs the host UUID of an agent from disk and then waits for the node with the given name to
+// show up in the cluster.
+//
+// The Electron app calls this method soon after starting the agent process.
+func (n *NodeJoinWait) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, cluster *clusters.Cluster) (clusters.Server, error) {
+	nodeName, err := n.getNodeNameFromHostUUIDFile(ctx, cluster)
+	if err != nil {
+		return clusters.Server{}, err
+	}
+
+	server, err := n.waitForNode(ctx, accessAndIdentity, cluster, nodeName)
+	if err != nil {
+		return clusters.Server{}, trace.Wrap(err)
+	}
+
+	return clusters.Server{
+		URI:    cluster.URI.AppendServer(server.GetName()),
+		Server: server,
+	}, nil
+}
+
+func (n *NodeJoinWait) getNodeNameFromHostUUIDFile(ctx context.Context, cluster *clusters.Cluster) (string, error) {
+	hostUUIDPath := filepath.Join(n.cfg.AgentsDir, cluster.ProfileName, "data", utils.HostUUIDFile)
+
+	// NodeJoinWait gets executed when the agent is booting up, so the host UUID file might not exist
+	// on disk yet. Use a ticker to periodically check for its existence.
+	ticker := time.NewTicker(50 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			out, err := utils.ReadPath(hostUUIDPath)
+			if err != nil {
+				if trace.IsNotFound(err) {
+					continue
+				}
+				return "", trace.Wrap(err)
+			}
+
+			id := strings.TrimSpace(string(out))
+			if id == "" {
+				return "", trace.NotFound("host UUID is empty")
+			}
+
+			return id, nil
```

Comment:
I couldn't use `utils.ReadHostUUID` because it returns `NotFound` both for when the file doesn't exist and when the host UUID in the file is empty.

I need to be able to distinguish between those two cases to enter the for loop again if the file was not created yet.

Reply:
Worth an inline comment